0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+  foreach left | left in s {
+    foreach right | right in s {
+      result := result + (left, right);
+    }
+  }
+}
+
+// Or even better, as an expression:
+function AllPairs(s: seq<nat>): seq<(nat, nat)> {
+  seq left, right | left in s && right in s
+}
+```
+
+A more serious shortcoming is that there is currently no efficient way to iterate over the elements of a `set` at all.
+The best alternative is to use the assign-such-that operator and set subtraction, illustrated in the reference manual as follows
+([sec 10.5.2](https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#1052-sets)):
+
+```dafny
+// s is a set<int>
+var ss := s;
+while ss != {}
+  decreases |ss|
+{
+  var i: int :| i in ss;
+  ss := ss - {i};
+  print i, "\n";
+}
+```
+
+This is functionally correct, but because set subtraction requires creating a fresh copy,
+when compiled this loop requires quadratic time and memory (or at least garbage-collection work) in the size of the set.
+The only way to avoid this cost is to write additional external code in the target language to interface directly with the
+internal runtime implementation of `set` values.
+
+By comparison, the equivalent `foreach` loop is:
+
+```dafny
+foreach i | i in s {
+  print i, "\n";
+}
+```
+
+The runtime implementation of this loop can use the native iteration features of the target language.
+
+Going further, it is better to avoid writing any imperative loop at all where possible. Sticking to immutable data and
+expressions rather than statements allows logic to be used in specifications as well as implementation,
+and reduces the burden on the Dafny verifier. The proposed sequence comprehension expression allows more
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in a set is simple using a sequence comprehension expression: `seq i | i in s`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## foreach Loops
+
+A `foreach` loop in Dafny resembles `for` or `foreach` loops in many other mainstream languages,
+but is substantially more general and expressive. A very simple example only looks slightly different than expected:
+
+```dafny
+foreach x | x in [1, 2, 3, 4, 5] {
+  print x, "\n";
+}
+```
+
+The feature supports much more sophisticated uses as well, however,
+including binding multiple variables at once and filtering:
+
+```dafny
+var myDoubleDeckerSeq: seq<seq<int>> := ...;
+foreach x, y | x in myDoubleDeckerSeq && y in x && y != 0 {
+  Process(y);
+}
+```
+
+The high-level syntax for a `foreach` loop reuses several existing concepts:
+
+```
+ForeachLoop ::=
+  "foreach"
+  QuantifierDomain
+  LoopSpec
+  Body
+```
+
+More specifically, this expands to:
+
+```
+ForeachLoop ::=
+  "foreach" 
+  Ident [":" Type] { "," Ident [":" Type] } [ "|" Expression ]
+  { InvariantDecl | ModifiesDecl | DecreasesDecl }
+  Body
```

Comment:
Would it be reasonable to consider that `while` loop is one of your `foreach` loops, under the following considerations:
- It's foreach loop with _no iteration variable_, not even one.
- The guard is the expression in the range
- Of course the _decreases clause_ is necessary.
?
If so, we might just want to replace the While loop internally by adding all these new iteration features.

Reply:
I don't think the semantics line up there correctly: if a loop with no bound variables was permitted, the semantics would be as I proposed further down, where the body would be executed once for every empty set of bindings produced in the domain's enumeration. That actually means such a `foreach` loop would be much closer to an `if` statement!

```dafny
var i := 0;
foreach | i <= 10 {
  // "i <= 10" is true, so the body would execute once
  print i, "\n";
  i := i + 1;
}
```

Your idea would only apply if you could quantify over the sequence of heap states somehow, I think.

You COULD translate `for` loops into `foreach` loops instead, but that will likely just make verifying the former cases more expensive, or at least add the burden of recognizing the special case of a domain where a single variable takes on those consecutive values. There was some discussion in the `for` loop RFC (#4) with the conclusion that there were benefits to having separate `for` and `foreach` loops, even if only to provide a simpler verification context for those learning Dafny.